UefiPayloadPkg/Library/CustomFdtNodeParserLib/CustomFdtNodeParserLib.c

```diff
@@ -0,0 +1,130 @@
+**/
+BOOLEAN
+FitIsHobNeed (
+  EFI_PEI_HOB_POINTERS  Hob
+  )
+{
+  if (FixedPcdGetBool (PcdHandOffFdtEnable)) {
+    if (Hob.Header->HobType == EFI_HOB_TYPE_HANDOFF) {
+      return FALSE;
+    }
+
+    if (Hob.Header->HobType == EFI_HOB_TYPE_MEMORY_ALLOCATION) {
+      if (CompareGuid (&Hob.MemoryAllocation->AllocDescriptor.Name, &gUniversalPayloadDeviceTreeGuid)) {
+        return FALSE;
+      }
+
+      if (CompareGuid (&Hob.MemoryAllocationModule->MemoryAllocationHeader.Name, &gEfiHobMemoryAllocModuleGuid)) {
+        return FALSE;
+      }
+
+      if ((Hob.MemoryAllocation->AllocDescriptor.MemoryType == EfiReservedMemoryType) ||
+          (Hob.MemoryAllocation->AllocDescriptor.MemoryType == EfiBootServicesCode) ||
+          (Hob.MemoryAllocation->AllocDescriptor.MemoryType == EfiBootServicesData) ||
+          (Hob.MemoryAllocation->AllocDescriptor.MemoryType == EfiRuntimeServicesCode) ||
+          (Hob.MemoryAllocation->AllocDescriptor.MemoryType == EfiRuntimeServicesData) ||
+          (Hob.MemoryAllocation->AllocDescriptor.MemoryType == EfiACPIReclaimMemory) ||
+          (Hob.MemoryAllocation->AllocDescriptor.MemoryType == EfiACPIMemoryNVS))
+      {
+        return FALSE;
+      }
+    }
+
+    if (Hob.Header->HobType == EFI_HOB_TYPE_GUID_EXTENSION) {
+      if (CompareGuid (&Hob.Guid->Name, &gUniversalPayloadSerialPortInfoGuid)) {
+        return FALSE;
+      }
+
+      if (CompareGuid (&Hob.Guid->Name, &gUniversalPayloadAcpiTableGuid)) {
+        return FALSE;
+      }
+    }
+  }
+
+  // Arrive here mean the HOB is need
+  return TRUE;
+}
+
+/**
+  It will Parse FDT -custom node based on information from bootloaders.
+  @param[in]  FdtBase The starting memory address of FdtBase
+  @param[in]  HostList The starting memory address of New Hob list.
+
+**/
+UINT64
+CustomFdtNodeParser (
+  IN VOID  *Fdt,
+  IN VOID  *HostList
+  )
+{
+  INT32               Node, CustomNode;
+  INT32               TempLen;
+  UINT64              *Data64;
+  UINT64              CHobList;
+  CONST FDT_PROPERTY  *PropertyPtr;
+
+  CHobList = 0;
+
+  DEBUG ((DEBUG_INFO, "%a() #1 \n", __func__));
+
+  //
+  // Look for if exists hob list node
+  //
+  Node = FdtSubnodeOffsetNameLen (Fdt, 0, "options", (INT32)AsciiStrLen ("options"));
+  if (Node > 0) {
+    DEBUG ((DEBUG_INFO, "  Found options node (%08X)", Node));
+    CustomNode = FdtSubnodeOffsetNameLen (Fdt, Node, "upl-custom", (INT32)AsciiStrLen ("upl-custom"));
+    if (CustomNode > 0) {
+      DEBUG ((DEBUG_INFO, "  Found upl-custom node (%08X)", CustomNode));
+      PropertyPtr = FdtGetProperty (Fdt, CustomNode, "hoblistptr", &TempLen);
+      Data64      = (UINT64 *)(PropertyPtr->Data);
+      CHobList    = Fdt64ToCpu (*Data64);
+      DEBUG ((DEBUG_INFO, "  Found hob list node (%08X)", CustomNode));
+      DEBUG ((DEBUG_INFO, " -pointer  %016lX\n", CHobList));
+    }
+  }
+
+  return CHobList;
+
+  DEBUG ((DEBUG_INFO, "%a() #3 \n", __func__));

```

Comment:
Are these debug prints left out or they are kept there intentionally?

Reply:
will remove it.